### PR TITLE
Update savgol_filter to include deriv functionality

### PR DESCRIPTION
**Pull request by jkv64**
This expands the functionality of the savgol_filter_dyn method to include the differentiation capability that the original [sci_py method](https://github.com/scipy/scipy/blob/main/scipy/signal/_savitzky_golay.py#L8-L144) has. This addition allows for the possibility of taking a derivative of a signal as it is convolved with the Savitsky-Golay filter. The [Wikipedia page](https://en.m.wikipedia.org/wiki/Savitzky%E2%80%93Golay_filter) has a good summary of the math involved, along with high quality visualizations.

This update also includes several more unit tests for the savgol_filter_dyn and savgol_coeffs_dyn methods and fixes an off by one error in the padding before convolution of savgol_filter_dyn.

A demonstration of different numbers of derivatives for fixed input is given below.
image

I'm looking for confirmation that everything seems documented properly, since this is my first addition to sci-rs. Let me know what you think @trueb2

### CODE REVIEW

File: sci-rs/src/signal/filter/savgol_filter.rs [hunk 1]
```diff
@@ -7,6 +7,12 @@ use nalgebra as na;
 use nalgebra::RealField;
 use num_traits::Float;
 
+//for testing
+#[cfg(feature = "std")]
+use std::fs::OpenOptions;
+#[cfg(feature = "std")]
+use std::io::Write;
+
 #[cfg(feature = "alloc")]
```
Comment (trueb2) [inline sci-rs/src/signal/filter/savgol_filter.rs:16]: Imports related to testing should only be included in the test mods. That would be moving imports into the 
```rust
#[cfg(test)]
mod tests {
  use super::*; //  just for import crate stuff, not test deps

  ...

}
```

Separately, I don't think we want tests that edit the file system, so these imports will be unnecessary.

File: sci-rs/src/signal/filter/savgol_filter.rs [hunk 2]
```diff
@@ -69,7 +82,12 @@ where
 /// This function is sensitive to f64 and f32 primitives due to use of least squares.
 /// The coefficients may go to zero for higher order polynomials and larger window lengths.
 ///
-pub fn savgol_coeffs_dyn<F>(window_length: usize, polyorder: usize) -> Vec<F>
+pub fn savgol_coeffs_dyn<F>(
+    window_length: usize,
+    polyorder: usize,
+    deriv: usize,
+    delta: F,
+) -> Vec<F>
```
Comment (trueb2) [inline sci-rs/src/signal/filter/savgol_filter.rs:90]: Here and for savgol_filter_dyn, we should accept parameters that make it clear what the defaults are like the kwargs. For example, deriv would be an Option<usize> that you would unwrap_or(0) and delta would be an Option<F> that you would unwrap or F::one()

File: sci-rs/src/signal/filter/savgol_filter.rs [hunk 3]
```diff
@@ -178,5 +402,64 @@ mod tests {
...
+        let data = (0..100).map(|i| (3 * i - 2) as f64); //y = 3x - 2
+        f.write_all(format!("{:?}\n", data.clone().collect::<Vec<_>>()).as_bytes())
+            .expect("Unable to write file");
+
+        for i in 0..3 {
+            let temp = data.clone();
+            let actual = savgol_filter_dyn(temp, 51, 5, i, 1.0);
+            f.write_all(format!("{:?}\n", actual).as_bytes())
+                .expect("Unable to write file");
+        }
```
Comment (trueb2) [inline sci-rs/src/signal/filter/savgol_filter.rs:463]: We should remove this test

**Review by trueb2 — CHANGES_REQUESTED** [review]
Thanks for the contribution. Looks like things are working but need to be cleaned up a bit.

**Comment by jkv64** [comment]
@trueb2 Thanks for the feedback! The changes should all be incorporated now.

**Comment by trueb2** [comment]
This looks good. Thanks for your contribution!

When this merges, the next release will need to bump the minor version since this includes some interface change.

**Review by trueb2 — APPROVED** [review]
